File: server/cache/extraction-cache.ts
/**
 * Extraction Result Caching Layer
 * Caches metadata extraction results to reduce redundant processing
 */

import { Redis } from 'ioredis';
import crypto from 'crypto';
import { CacheConfig, DEFAULT_CACHE_CONFIG } from './cache-config';

export interface ExtractionCacheConfig extends CacheConfig {
  /** TTL for extraction results in seconds */
  extractionTTL: number;
  /** TTL for field-specific results in seconds */
  fieldTTL: number;
  /** TTL for tier-filtered results in seconds */
  tierTTL: number;
  /** Maximum size of cached extraction results in MB */
  maxResultSize: number;
}

export const DEFAULT_EXTRACTION_CACHE_CONFIG: ExtractionCacheConfig = {
  ...DEFAULT_CACHE_CONFIG,
  extractionTTL: 3600, // 1 hour for extraction results
  fieldTTL: 1800,      // 30 minutes for field-specific results
  tierTTL: 7200,       // 2 hours for tier-filtered results
  maxResultSize: 10,   // 10MB max result size
};

export interface ExtractionCacheEntry {
  key: string;
  value: any;
  timestamp: number;
  ttl: number;
  metadata: {
    fileType: string;
    tier: string;
    fieldsExtracted: number;
    processingTime: number;
    size: number;
  };
}

export class ExtractionCache {
  private redis: Redis;
  private config: ExtractionCacheConfig;

  constructor(redis: Redis, config?: Partial<ExtractionCacheConfig>) {
    this.redis = redis;
    this.config = { ...DEFAULT_EXTRACTION_CACHE_CONFIG, ...config };
  }

  /**
   * Generate a cache key for extraction results
   */
  generateExtractionKey(filePath: string, tier: string, options: any = {}): string {
    const fileHash = crypto.createHash('sha256').update(filePath).digest('hex');
    const optionsHash = crypto.createHash('sha256').update(JSON.stringify(options)).digest('hex');
    
    return `extraction:${fileHash}:${tier}:${optionsHash}`;
  }

  /**
   * Generate a cache key for field-specific results
   */
  generateFieldKey(filePath: string, tier: string, field: string): string {
    const fileHash = crypto.createHash('sha256').update(filePath).digest('hex');
    
    return `field:${fileHash}:${tier}:${field}`;
  }

  /**
   * Cache extraction results
   */
  async setExtractionResult(
    filePath: string, 
    tier: string, 
    result: any, 
    options: any = {}
  ): Promise<boolean> {
    try {
      const key = this.generateExtractionKey(filePath, tier, options);
      const size = JSON.stringify(result).length;
      
      // Don't cache if result is too large
      if (size > this.config.maxResultSize * 1024 * 1024) {
        console.log(`Extraction result too large to cache: ${size} bytes`);
        return false;
      }

      const entry: ExtractionCacheEntry = {
        key,
        value: result,
        timestamp: Date.now(),
        ttl: this.config.extractionTTL,
        metadata: {
          fileType: options.fileType || 'unknown',
          tier,
          fieldsExtracted: result.fieldsExtracted || 0,
          processingTime: result.processingTime || 0,
          size,
        }
      };

      await this.redis.setex(key, this.config.extractionTTL, JSON.stringify(entry));
      console.log(`Cached extraction result for ${filePath} (tier: ${tier})`);
      return true;
    } catch (error) {
      console.error('Error caching extraction result:', error);
      return false;
    }
  }

  /**
   * Get cached extraction results
   */
  async getExtractionResult(
    filePath: string, 
    tier: string, 
    options: any = {}
  ): Promise<any | null> {
    try {
      const key = this.generateExtractionKey(filePath, tier, options);
      const cached = await this.redis.get(key);
      
      if (!cached) {
        return null;
      }

      const entry: ExtractionCacheEntry = JSON.parse(cached);
      console.log(`Cache hit for extraction result: ${filePath} (tier: ${tier})`);
      return entry.value;
    } catch (error) {
      console.error('Error getting cached extraction result:', error);
      return null;
    }
  }

  /**
   * Cache a specific field result
   */
  async setFieldResult(
    filePath: string,
    tier: string,
    field: string,
    result: any
  ): Promise<boolean> {
    try {
      const key = this.generateFieldKey(filePath, tier, field);
      const entry: ExtractionCacheEntry = {
        key,
        value: result,
        timestamp: Date.now(),
        ttl: this.config.fieldTTL,
        metadata: {
          fileType: 'unknown', // Would need to be passed in
          tier,
          fieldsExtracted: 1,
          processingTime: 0,
          size: JSON.stringify(result).length,
        }
      };

      await this.redis.setex(key, this.config.fieldTTL, JSON.stringify(entry));
      console.log(`Cached field result for ${filePath} (field: ${field})`);
      return true;
    } catch (error) {
      console.error('Error caching field result:', error);
      return false;
    }
  }

  /**
   * Get cached field result
   */
  async getFieldResult(
    filePath: string,
    tier: string,
    field: string
  ): Promise<any | null> {
    try {
      const key = this.generateFieldKey(filePath, tier, field);
      const cached = await this.redis.get(key);
      
      if (!cached) {
        return null;
      }

      const entry: ExtractionCacheEntry = JSON.parse(cached);
      console.log(`Cache hit for field result: ${filePath} (field: ${field})`);
      return entry.value;
    } catch (error) {
      console.error('Error getting cached field result:', error);
      return null;
    }
  }

  /**
   * Cache tier-specific filtered results
   */
  async setTierFilteredResult(
    filePath: string,
    originalTier: string,
    targetTier: string,
    result: any
  ): Promise<boolean> {
    try {
      const key = `tier-filter:${filePath}:${originalTier}:${targetTier}`;
      const entry: ExtractionCacheEntry = {
        key,
        value: result,
        timestamp: Date.now(),
        ttl: this.config.tierTTL,
        metadata: {
          fileType: 'unknown',
          tier: targetTier,
          fieldsExtracted: result.fieldsExtracted || 0,
          processingTime: 0,
          size: JSON.stringify(result).length,
        }
      };

      await this.redis.setex(key, this.config.tierTTL, JSON.stringify(entry));
      console.log(`Cached tier-filtered result for ${filePath} (${originalTier} -> ${targetTier})`);
      return true;
    } catch (error) {
      console.error('Error caching tier-filtered result:', error);
      return false;
    }
  }

  /**
   * Get cached tier-specific filtered results
   */
  async getTierFilteredResult(
    filePath: string,
    originalTier: string,
    targetTier: string
  ): Promise<any | null> {
    try {
      const key = `tier-filter:${filePath}:${originalTier}:${targetTier}`;
      const cached = await this.redis.get(key);
      
      if (!cached) {
        return null;
      }

      const entry: ExtractionCacheEntry = JSON.parse(cached);
      console.log(`Cache hit for tier-filtered result: ${filePath} (${originalTier} -> ${targetTier})`);
      return entry.value;
    } catch (error) {
      console.error('Error getting cached tier-filtered result:', error);
      return null;
    }
  }

  /**
   * Invalidate extraction cache for a specific file
   */
  async invalidateExtractionCache(filePath: string): Promise<number> {
    try {
      const fileHash = crypto.createHash('sha256').update(filePath).digest('hex');
      const pattern = `extraction:${fileHash}:*`;
      
      // Get all keys matching the pattern
      const keys = await this.redis.keys(pattern);
      
      if (keys.length === 0) {
        return 0;
      }

      // Delete all matching keys
      const result = await this.redis.del(...keys);
      console.log(`Invalidated ${result} extraction cache entries for ${filePath}`);
      return result;
    } catch (error) {
      console.error('Error invalidating extraction cache:', error);
      return 0;
    }
  }

  /**
   * Get cache statistics
   */
  async getStats(): Promise<any> {
    try {
      const info = await this.redis.info('keyspace');
      return {
        extractionCacheEnabled: this.config.enabled,
        extractionTTL: this.config.extractionTTL,
        fieldTTL: this.config.fieldTTL,
        tierTTL: this.config.tierTTL,
        maxResultSize: this.config.maxResultSize,
        redisInfo: info,
      };
    } catch (error) {
      console.error('Error getting cache stats:', error);
      return null;
    }
  }
}

File: server/cache/extraction_cache.py
"""
Extraction Result Cache for MetaExtract

Caches complete metadata extraction results to avoid re-running expensive operations.
"""

import logging
import time
from typing import Any, Dict, Optional
from datetime import timedelta

from .base_cache import BaseCache

logger = logging.getLogger("metaextract.cache.extraction")


class ExtractionCache(BaseCache):
    """
    Caches complete metadata extraction results.
    
    Features:
    - File content-based cache invalidation
    - Tier-specific caching
    - Format-specific optimization
    - Processing time tracking
    """
    
    def __init__(self):
        """Initialize extraction cache with 1-hour TTL."""
        super().__init__(cache_prefix="extraction", default_ttl=3600)  # 1 hour
    
    def get_cache_key(self, file_path: str, tier: str = "free", 
                     file_format: Optional[str] = None) -> str:
        """
        Generate cache key for extraction result.
        
        Args:
            file_path: Path to the file
            tier: User tier level
            file_format: File format (optional)
            
        Returns:
            Cache key string
        """
        # Calculate file hash for content-based invalidation
        file_hash = self._calculate_file_hash(file_path, quick=True)
        
        # Include tier and format in key
        key_components = [file_hash, tier]
        if file_format:
            key_components.append(file_format)
        
        return self._generate_cache_key(*key_components)
    
    def get_cached_result(self, file_path: str, tier: str = "free",
                         file_format: Optional[str] = None) -> Optional[Dict[str, Any]]:
        """
        Get cached extraction result if available and valid.
        
        Args:
            file_path: Path to the file
            tier: User tier level
            file_format: File format (optional)
            
        Returns:
            Cached metadata or None if not available/valid
        """
        cache_key = self.get_cache_key(file_path, tier, file_format)
        
        # Check if file has been modified
        cached_result = self.get(cache_key)
        if cached_result is None:
            return None
        
        # Validate cached result structure
        if not isinstance(cached_result, dict):
            logger.warning(f"Invalid cached result format for {file_path}")
            self.delete(cache_key)
            return None
        
        # Check if cached file hash matches current file
        current_file_hash = self._calculate_file_hash(file_path, quick=True)
        cached_file_hash = cached_result.get('file_hash')
        
        if cached_file_hash != current_file_hash:
            logger.debug(f"File {file_path} modified, invalidating cache")
            self.delete(cache_key)
            return None
        
        # Add cache hit information
        cached_result['cache_info'] = {
            'hit': True,
            'cached_at': cached_result.get('cached_at'),
            'cache_key': cache_key
        }
        
        logger.debug(f"Cache hit for extraction: {file_path} (tier: {tier})")
        return cached_result
    
    def cache_result(self, result: Dict[str, Any], file_path: str, 
                    tier: str = "free", file_format: Optional[str] = None,
                    processing_time_ms: Optional[float] = None) -> bool:
        """
        Cache extraction result.
        
        Args:
            result: Extraction result to cache
            file_path: Path to the file
            tier: User tier level
            file_format: File format (optional)
            processing_time_ms: Processing time in milliseconds
            
        Returns:
            True if caching was successful
        """
        cache_key = self.get_cache_key(file_path, tier, file_format)
        
        # Add metadata for cache validation
        cache_data = result.copy()
        cache_data['file_hash'] = self._calculate_file_hash(file_path, quick=True)
        cache_data['cached_at'] = int(time.time())
        cache_data['tier'] = tier
        cache_data['file_format'] = file_format
        cache_data['processing_time_ms'] = processing_time_ms
        
        # Remove any existing cache info to avoid confusion
        cache_data.pop('cache_info', None)
        
        # Set appropriate TTL based on tier
        if tier == "super":
            ttl = 7200  # 2 hours for super tier
        elif tier == "premium":
            ttl = 5400  # 1.5 hours for premium tier
        else:
            ttl = self.default_ttl  # Default 1 hour
        
        success = self.set(cache_key, cache_data, ttl=ttl)
        
        if success:
            logger.debug(f"Cached extraction result: {file_path} (tier: {tier}, TTL: {ttl}s)")
        else:
            logger.warning(f"Failed to cache extraction result: {file_path}")
        
        return success
    
    def invalidate_file(self, file_path: str, tier: Optional[str] = None) -> int:
        """
        Invalidate all cached results for a file.
        
        Args:
            file_path: Path to the file
            tier: Specific tier to invalidate (optional)
            
        Returns:
            Number of invalidated entries
        """
        if tier:
            # Generate the exact cache key for this file+tier combination
            cache_key = self.get_cache_key(file_path, tier)
            success = self.delete(cache_key)
            count = 1 if success else 0
        else:
            # Invalidate all tiers - need to check all possible tiers
            count = 0
            for tier_name in ["free", "starter", "premium", "super"]:
                cache_key = self.get_cache_key(file_path, tier_name)
                if self.exists(cache_key):
                    if self.delete(cache_key):
                        count += 1
        logger.info(f"Invalidated {count} extraction cache entries for {file_path}")
        return count
    
    def get_tier_stats(self, tier: str) -> Dict[str, Any]:
        """Get cache statistics for specific tier."""
        stats = self.get_stats()
        
        # Count keys for this tier
        tier_pattern = f"*_{tier}*"
        tier_keys = self.redis_client.keys(f"{self.cache_prefix}:{tier_pattern}")
        
        stats['tier_keys'] = len(tier_keys)
        stats['tier'] = tier
        
        return stats
    
    def warm_cache(self, file_paths: list, tier: str = "free") -> Dict[str, bool]:
        """
        Pre-warm cache by extracting metadata for multiple files.
        
        Args:
            file_paths: List of file paths to warm
            tier: User tier level
            
        Returns:
            Dictionary mapping file paths to success status
        """
        from ..extractor.core.comprehensive_engine import NewComprehensiveMetadataExtractor
        
        results = {}
        extractor = NewComprehensiveMetadataExtractor()
        
        for file_path in file_paths:
            try:
                # Check if already cached
                if self.get_cached_result(file_path, tier) is not None:
                    results[file_path] = True
                    continue
                
                # Extract and cache
                start_time = time.time()
                result = extractor.extract_comprehensive_metadata(file_path, tier)
                processing_time = (time.time() - start_time) * 1000
                
                # Cache the result
                success = self.cache_result(
                    result, file_path, tier, 
                    processing_time_ms=processing_time
                )
                
                results[file_path] = success
                
            except Exception as e:
                logger.error(f"Failed to warm cache for {file_path}: {e}")
                results[file_path] = False
        
        logger.info(f"Cache warming completed: {sum(results.values())}/{len(results)} files")
        return results